openlr/openlr_decoder.cpp

```diff
@@ -176,6 +185,169 @@ void CopyWithoutOffsets(InputIterator const start, InputIterator const stop, Out
+    for (auto const & point : ref.m_points)
+      m_points.emplace_back(point);
+
+    auto positiveOffsetM = ref.m_positiveOffsetMeters;
+    if (positiveOffsetM >= m_points[0].m_distanceToNextPointM)
+    {
+      LOG(LWARNING, ("Wrong positive offset for segment:", segment.m_segmentId));
+      positiveOffsetM = 0;
+    }
+
+    auto negativeOffsetM = ref.m_negativeOffsetMeters;
+    if (negativeOffsetM >= m_points[m_points.size() - 2].m_distanceToNextPointM)
+    {
+      LOG(LWARNING, ("Wrong negative offset for segment:", segment.m_segmentId));
+      negativeOffsetM = 0;
+    }
+
+    {
+      double expectedLength = 0;
+      for (size_t i = 0; i + 1 < m_points.size(); ++i)
+        expectedLength += m_points[i].m_distanceToNextPointM;
+
+      if (positiveOffsetM + negativeOffsetM >= expectedLength)
+      {
+        LOG(LINFO, ("Skipping", segment.m_segmentId, "due to wrong positive/negative offsets."));
+        return false;
+      }
+
+      if (positiveOffsetM + negativeOffsetM + kOffsetToleranceM >= expectedLength)
+      {
+        LOG(LINFO, ("Too tight positive and negative offsets, setting them to zero."));
+        positiveOffsetM = 0;
+        negativeOffsetM = 0;
+        ++stat.m_tightOffsets;
+      }
+    }
+
+    if (!m_router.Go(m_points, positiveOffsetM, negativeOffsetM, path.m_path))
+      return false;
+
+    return true;
+  }
+
+private:
+  FeaturesRoadGraph m_roadGraph;
+  RoadInfoGetter m_infoGetter;
+  Router m_router;
+  vector<WayPoint> m_points;
+};
+
+class SegmentsDecoderV2
+{
+public:
+  SegmentsDecoderV2(Index const & index, unique_ptr<CarModelFactory> cmf)
+    : m_index(index), m_graph(index, move(cmf)), m_infoGetter(index)
+  {
+  }
+
+  bool DecodeSegment(LinearSegment const & segment, DecodedPath & path, v2::Stats & stat)
+  {
+    double const kPathLengthTolerance = 0.30;
+    uint32_t const kMaxJunctionCandidates = 10;
+    uint32_t const kMaxProjectionCandidates = 5;
+
+    m_graph.ResetFakes();
+
+    path.m_segmentId.Set(segment.m_segmentId);
+
+    auto const & points = segment.GetLRPs();
+    vector<vector<Graph::EdgeVector>> lineCandidates;
+    lineCandidates.reserve(points.size());
+    LOG(LDEBUG, ("Decoding segment:", segment.m_segmentId, "with", points.size(), "points"));
+
+    CandidatePointsGetter pointsGetter(kMaxJunctionCandidates, kMaxProjectionCandidates, m_index,
+                                       m_graph);
+    CandidatePathsGetter pathsGetter(pointsGetter, m_graph, m_infoGetter, stat);
+
+    if (!pathsGetter.GetLineCandidatesForPoints(points, lineCandidates))
+      return false;
+
+    vector<Graph::EdgeVector> resultPath;
+    PathsConnector connector(kPathLengthTolerance, m_graph, stat);
+    if (!connector.ConnectCandidates(points, lineCandidates, resultPath))
+      return false;
+
+    Graph::EdgeVector route;
+    for (auto const & part : resultPath)
+      route.insert(end(route), begin(part), end(part));
+
+    double requiredRouteDistanceM = 0.0;
```

Comment:
Please, add a CHECK that points are not empty.

Reply:
Done,